src/persist-client/examples/persistent_source.rs

```diff
@@ -0,0 +1,568 @@
+                        }
+
+                        // The "real" timestamper would have to use a compare-and-set mechanism to write down
+                        // new bindings in some system backed by consensus.
+                        let binding = match binding {
+                            Some(binding) => binding,
+                            None => {
+                                let now = (self.now_fn)();
+                                let now_clamped =
+                                    now - (now % self.timestamp_interval.as_millis() as u64);
+                                let mut current_ts = self.current_ts.lock().await;
+                                if now_clamped != *current_ts {
+                                    *current_ts = now_clamped;
+                                }
+
+                                let new_binding = (*upper_element, *current_ts);
+                                shared_bindings.push(new_binding);
+                                new_binding
+                            }
+                        };
+
+                        result_bindings.push(binding);
+                        e.insert(binding);
+                    }
+                }
+            }
+
+            TimestampBindings {
+                bindings: result_bindings,
+                lower,
+                upper,
+            }
+        }
+    }
+}
+
+mod render {
+    use std::error::Error;
+    use std::fmt::Debug;
+    use std::time::Duration;
+
+    use tokio::sync::mpsc;
+    use tracing::error;
+
+    use crate::api::{Batch, Source, TimestampedBatch, Timestamper};
+
+    /// Spawns tokio tasks for the different steps of a source pipeline. We read from the source,
+    /// decode, assign timestamps to updates, and emit them.
+    pub async fn spawn_source_pipeline<S, TS, T>(
+        name: String,
+        source: S,
+        timestamper: TS,
+        batch_size: usize,
+    ) -> Result<(), Box<dyn Error>>
+    where
+        S: Source<SourceTimestamp = T> + Send + 'static,
+        TS: Timestamper<SourceTimestamp = T> + Send + Sync + 'static,
+        T: timely::PartialOrder + PartialOrd + Copy + Debug + Send + Sync + 'static,
+        S::SourceTimestamp: Send,
+        S::D: Send,
+        TS::SourceTimestamp: Send,
+        TS::AssignedTimestamp: Send,
+    {
+        // Right now, everything is a separate task, including the prefetching. We'd have to see
+        // how to split these in a real implementation.
+
+        let (raw_batches_tx, raw_batches_rx) = mpsc::channel(2);
+
+        let source_task = tokio::task::spawn(async move {
+            read_source(source, batch_size, raw_batches_tx).await;
+        });
+
+        let (raw_batches_tx2, raw_batches_rx2) = mpsc::channel(2);
+
+        let timestamper1 = timestamper.clone();
+        let prefetcher_task = tokio::task::spawn(async move {
+            prefetch_timestamp_bindings(timestamper1, raw_batches_rx, raw_batches_tx2).await;
```

Comment:
do we think the real system will be routing all the source data through the timestamper or would the real source have an API more like 1) notify me of new data which announces some sort of token/index (for the timestamper) 2) fetch new data based the token/index having been routed through timestamping?

it's 100% fine if this example is simplified for illustrative purposes, but if so, let's leave a comment to that effect

Reply:
It's more for illustrative purposes but I don't yet have an idea on how the real thing would work. The source could just, when you read a batch, also return the offset uppers, and we notify the timestamping logic so that it can start minting early.